drivers/meter/ade7754/ade7754.c

```diff
@@ -0,0 +1,1156 @@
+
+	switch (sel) {
+	case ADE7754_PHASE_C:
+		ret = ade7754_update_bits(dev, ADE7754_REG_MMODE, ADE7754_PEAKSEL_MSK,
+					  no_os_field_prep(ADE7754_PEAKSEL_MSK, ADE7754_PHASE_C));
+		break;
+	case ADE7754_PHASE_B:
+		ret = ade7754_update_bits(dev, ADE7754_REG_MMODE, ADE7754_PEAKSEL_MSK,
+					  no_os_field_prep(ADE7754_PEAKSEL_MSK, ADE7754_PHASE_B));
+		break;
+	default:
+		ret = ade7754_update_bits(dev, ADE7754_REG_MMODE, ADE7754_PEAKSEL_MSK,
+					  no_os_field_prep(ADE7754_PEAKSEL_MSK, ADE7754_PHASE_A));
+		break;
+	}
+
+	return ret;
+}
+
+/**
+ * @brief Operation mode of the ADCs.
+ * @param dev - The device structure.
+ * @param sel - mode of the ADCs
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ade7754_adcs_mode(struct ade7754_dev *dev,
+		      enum ade7754_dismod sel)
+{
+	int ret;
+
+	switch (sel) {
+	case ADE7754_POWER_DOWN_MODE:
+		ret = ade7754_update_bits(dev, ADE7754_REG_OPMODE, ADE7754_DISMOD_MSK,
+					  no_os_field_prep(ADE7754_DISMOD_MSK, ADE7754_POWER_DOWN_MODE));
+		break;
+	case ADE7754_OFF_VOLTAGE_ADCS_REDIRECTED_TO_CURRENT:
+		ret = ade7754_update_bits(dev, ADE7754_REG_OPMODE, ADE7754_DISMOD_MSK,
+					  no_os_field_prep(ADE7754_DISMOD_MSK,
+							  ADE7754_OFF_VOLTAGE_ADCS_REDIRECTED_TO_CURRENT));
+		break;
+	case ADE7754_OFF_CURRENT_ADCS_REDIRECTED_TO_VOLTAGE:
+		ret = ade7754_update_bits(dev, ADE7754_REG_OPMODE, ADE7754_DISMOD_MSK,
+					  no_os_field_prep(ADE7754_DISMOD_MSK,
+							  ADE7754_OFF_CURRENT_ADCS_REDIRECTED_TO_VOLTAGE));
+		break;
+	case ADE7754_SWAP_VOLTAGE_CURRENT:
+		ret = ade7754_update_bits(dev, ADE7754_REG_OPMODE, ADE7754_DISMOD_MSK,
+					  no_os_field_prep(ADE7754_DISMOD_MSK, ADE7754_SWAP_VOLTAGE_CURRENT));
+		break;
+	case ADE7754_SLEEP_MODE:
+		ret = ade7754_update_bits(dev, ADE7754_REG_OPMODE, ADE7754_DISMOD_MSK,
+					  no_os_field_prep(ADE7754_DISMOD_MSK, ADE7754_SLEEP_MODE));
+		break;
+	case ADE7754_SWITCH_OFF_VOLTAGE_ADC:
+		ret = ade7754_update_bits(dev, ADE7754_REG_OPMODE, ADE7754_DISMOD_MSK,
+					  no_os_field_prep(ADE7754_DISMOD_MSK, ADE7754_SWITCH_OFF_VOLTAGE_ADC));
+		break;
+	case ADE7754_SWITCH_OFF_CURRENT_ADC:
+		ret = ade7754_update_bits(dev, ADE7754_REG_OPMODE, ADE7754_DISMOD_MSK,
+					  no_os_field_prep(ADE7754_DISMOD_MSK, ADE7754_SWITCH_OFF_CURRENT_ADC));
+		break;
+	default:
+		ret = ade7754_update_bits(dev, ADE7754_REG_OPMODE, ADE7754_DISMOD_MSK,
+					  no_os_field_prep(ADE7754_DISMOD_MSK, ADE7754_NORMAL_OPERATION));
+		break;
+	}
+
+	return ret;
+}
+
+/**
+ * @brief Read energy values
+ * @param dev - The device structure.
+ * @param data - Structure to store energy values
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ade7754_energy_vals(struct ade7754_dev *dev,
+			struct ade7754_energy_values *data)
+{
+	int ret;
+	/* temporary value read from register */
+	uint32_t temp_val;
+
+	if (!data)
+		return -EINVAL;
+
+	/* read accumulated active energy */
+	ret = ade7754_read(dev, ADE7754_REG_AENERGY, &temp_val);
```

Comment:
hmm, i am a bit confused, you use temp_val (uint32_t) as input to a function that expects int32_t, then you cast it to int32_t below. Why not making temp_val int32_t from the start?